sklearn/utils/tests/test_stats.py

```diff
@@ -12,121 +12,151 @@
 
-    Ensures that (leading)zero-weight observations ignored when `percentile_rank=0`.
+    Check that leading zero-weight observations ignored when `percentile_rank=0`.
     See #20528 for details.
+    Check that when `average=True` and the `j+1` ('plus one') index has sample weight
+    of 0, it is ignored. Also check that trailing zero weight observations are ignored
+    (e.g., when `percentile_rank=100`).
     """
-    y = np.array([0, 1, 2, 3, 4, 5])
-    sw = np.array([0, 0, 1, 1, 1, 0])
-    value = _weighted_percentile(y, sw, 0)
-    assert approx(value) == 2
-
-    value = _weighted_percentile(y, sw, 50)
-    assert approx(value) == 3
-
-    value = _weighted_percentile(y, sw, 100)
-    assert approx(value) == 4
+    y = np.array([0, 1, 2, 3, 4, 5, 6])
+    sw = np.array([0, 0, 1, 1, 0, 1, 0])
 
-
-def test_weighted_median_equal_weights(global_random_seed):
-    """Checks `_weighted_percentile(percentile_rank=50)` is the same as `np.median`.
-
-    `sample_weights` are all 1s and the number of samples is odd.
-    When number of samples is odd, `_weighted_percentile` always falls on a single
-    observation (not between 2 values, in which case the lower value would be taken)
-    and is thus equal to `np.median`.
-    For an even number of samples, this check will not always hold as (note that
-    for some other percentile methods it will always hold). See #17370 for details.
-    """
-    rng = np.random.RandomState(global_random_seed)
-    x = rng.randint(10, size=11)
-    weights = np.ones(x.shape)
-    median = np.median(x)
-    w_median = _weighted_percentile(x, weights)
-    assert median == approx(w_median)
+    value = _weighted_percentile(
+        np.vstack((y, y)).T, np.vstack((sw, sw)).T, percentile_rank, average=average
+    )
+    for idx in range(2):
+        assert approx(value[idx]) == expected_value
 
 
-def test_weighted_median_integer_weights(global_random_seed):
-    # Checks average weighted percentile_rank=0.5 is same as median when manually weight
-    # data
+@pytest.mark.parametrize("average", [True, False])
+@pytest.mark.parametrize("percentile_rank", [20, 35, 61])
+def test_weighted_median_frequency_weights(
+    global_random_seed, percentile_rank, average
+):
+    """Check integer weights give the same result as repeating values."""
     rng = np.random.RandomState(global_random_seed)
     x = rng.randint(20, size=10)
     weights = rng.choice(5, size=10)
-    x_manual = np.repeat(x, weights)
-    median = np.median(x_manual)
-    w_median = _averaged_weighted_percentile(x, weights)
-    assert median == approx(w_median)
 
+    x_repeated = np.repeat(x, weights)
+    percentile_weights = _weighted_percentile(
+        x, weights, percentile_rank, average=average
+    )
+    percentile_repeated = _weighted_percentile(
+        x_repeated, np.ones_like(x_repeated), percentile_rank, average=average
+    )
+    assert percentile_weights == approx(percentile_repeated)
```

Comment:
Please (re-)include the `percentile_rank=50` case and checks that it also matches the results of `np.median(x_repeated)` when `average is True`.

Reply:
```suggestion
    assert percentile_weights == approx(percentile_repeated)

    if percentile_rank == 50 and average:
        assert percentile_weights == approx(np.median(x_repeated))
```

Reply:
Done